### PR TITLE
Improve --help performance for x.py

### LINKED ISSUE #37305: bootstrap.py --help starts with downloading
rustbuild really has a proper --help response, but the first time the user asks for it, it starts working / downloading right away. This is a bad first impression, and a stumbling block for getting to know the new build system.

Another usability issue:

```
$ python src/bootstrap/bootstrap.py --help
    Blocking waiting for file lock on the git checkouts
```

### PR DESCRIPTION
**Pull request by Phlosioneer**
Since compiling the bootstrap command doesn't require any submodules,
we can skip updating submodules when a --help command is passed in.
On my machine, this saves 1 minute if the submodules are already
downloaded, and 10 minutes if run on a clean repo.

This commit also adds a message before compiling/downloading anything
when a --help command is passed in, to tell the user WHY --help
takes so long to complete. It also points the user to the bootstrap
README.md for faster help.

Finally, this fixes one warning message that still referenced using
make instead of x.py, even though x.py is now the standard way of
building rust.

Closes #37305

### CODE REVIEW

File: src/bootstrap/bootstrap.py [hunk 1]
```diff
@@ -760,7 +763,13 @@ def main():
     help_triggered = (
         '-h' in sys.argv) or ('--help' in sys.argv) or (len(sys.argv) == 1)
     try:
-        bootstrap()
+        # If the user is asking for help, let them know that the whole download-and-build
+        # process has to happen before anything is printed out.
+        if help_triggered:
+            print("NOTE: Downloading and compiling bootstrap requirements before processing")
+            print("      --help command. See src/bootstrap/README.md for help with common")
+            print("      commands.")
```
Comment (petrochenkov) [inline src/bootstrap/bootstrap.py]: We should get this message only if we actually going to download something or do other long operation.

Also there's already `print_what_bootstrap_means` that is supposed to print this exact information, but maybe it doesn't fire for some reason?

Comment (Phlosioneer) [reply src/bootstrap/bootstrap.py]: I decided to print the message every time, because we _always_ rebuild the build tool. So it will _never_ be instant. Also, doing anything more complicated than printing this at the beginning means that we have to go through the whole process to gather information, setup environment variables, and check if anything needs to be done. It would require a lot of logic spread through a lot of the build script, and that complexity simply isn't worth it.

I have never noticed `print_what_bootstrap_means`. Looking back at it, it only triggers if a download happens, and is _immediately_ followed by a lot of download messages and loading bars. Also, prior to this fix, the very first thing the build script did was `update_submodules`, so `print_what_bootstrap_means` would be printed after the most time consuming part of the build script.

The message also doesn't reference --help at all; it states why stuff needs to download before running build commands, but the user will not expect --help to be part of that. My message makes that perfectly clear.

Basically, we either have to print this whenever --help happens, regardless of what we actually need to do; OR we have to wait until the build script sets up enough to _know_ if it needs to download, at which point seconds have already passed and the tool will feel sluggish (asking the user to wait after the user has already waited). And if we want to do the second option, because of how the script is organized, we would either have to do a lot of refactoring, or litter the code with checks and print statements.

Comment (petrochenkov) [reply src/bootstrap/bootstrap.py]: Ok, sounds good.
Could you remove `print_what_bootstrap_means` (and also `self.printed`) then?

File: src/bootstrap/bootstrap.py [hunk 2]
```diff
@@ -760,7 +763,13 @@ def main():
     help_triggered = (
         '-h' in sys.argv) or ('--help' in sys.argv) or (len(sys.argv) == 1)
     try:
-        bootstrap()
+        # If the user is asking for help, let them know that the whole download-and-build
+        # process has to happen before anything is printed out.
+        if help_triggered:
```
Comment (petrochenkov) [inline src/bootstrap/bootstrap.py]: Could you move this into `bootstrap()`? Top level `main` is for catching and handling exceptions only.

File: src/bootstrap/bootstrap.py [hunk 3]
```diff
@@ -760,7 +763,13 @@ def main():
...
+            print("NOTE: Downloading and compiling bootstrap requirements before processing")
```
Comment (petrochenkov) [inline src/bootstrap/bootstrap.py]: Is the word "requirements" used in this meaning? "Downloading requirements" sounds weird.
Maybe "Downloading and compiling bootstrap and its dependencies ..."?

**Comment by petrochenkov** [comment]
Thanks!
@bors r+
